### ISSUE TITLE
Safari 10 detection for indexedDB fails

### ISSUE THREAD

**Issue by willmcclellan** [localForage/localForage#604]
(Desktop) Safari 9.1 detection for indexedDB seems to fail at this conditional:

```
        // We mimic PouchDB here; just UA test for Safari (which, as of
        // iOS 8/Yosemite, doesn't properly support IndexedDB).
        // IndexedDB support is broken and different from Blink's.
        // This is faster than the test case (and it's sync), so we just
        // do this. *SIGH*
        // http://bl.ocks.org/nolanlawson/raw/c83e9039edf2278047e9/
        //
        // We test for openDatabase because IE Mobile identifies itself
        // as Safari. Oh the lulz...
        if (typeof openDatabase !== 'undefined' && typeof navigator !== 'undefined' &&
            navigator.userAgent &&
            /Safari/.test(navigator.userAgent) && !/Chrome/.test(navigator.userAgent)) {
            return false;
        }
```

https://github.com/localForage/localForage/blob/master/src/utils/isIndexedDBValid.js#L19-L23

IndexedDB does work, but because it doesn't pass this conditional it falls back to webSQL which doesn't seem to work in my setup.

**Comment by paulkok** [comment]
IndexedDB is broken on iOS 8 & 9, see the comments and link (http://bl.ocks.org/nolanlawson/raw/c83e9039edf2278047e9/) in the code. 

But I confirmed IndexedDB IS working fine on iOS 10 so we should exclude Safari on iOS 10+ from this check.

webSQL on Safari on iOS 8&9 is working by the way but it does have the 50MB limit etc.

**Comment by thgreasi** [comment]
Also [caniuse](http://caniuse.com/#feat=indexeddb) doesn't (yet) have any compatibility notes,
@nolanlawson have you yet handled it in PouchDB side?

**Comment by nolanlawson** [comment]
@thgreasi based on my tests, the WebKit team has fixed all outstanding bugs in IndexedDB in the latest release. I've been hesitant to switch PouchDB yet because it also seems much slower than WebSQL (~30x or more), so I worry that if we switch people will just struggle to put it back on IndexedDB because their app slowed down.

It needs more investigation, though. I haven't had time to do a thorough benchmark of WebSQL vs IndexedDB.

**Comment by nolanlawson** [comment]
Related: https://github.com/pouchdb/pouchdb/issues/5572

**Comment by willmcclellan** [comment]
I should add that when our app falls back to WebSQL it errors on Safari 9.1 with the error:

`self.ready is not a function. (In 'self.ready()', 'self.ready' is undefined)` 

From this line: https://github.com/localForage/localForage/blob/master/src/drivers/websql.js#L151

**Comment by Narcwis** [comment]
I have tried both indexedDB and WebSQL (by using the config() method) but none seem to work. The console shows data being set but when getItem is called the data is blank {}. Storage panel (Web Inspector) only confirms that the db is in fact created but the records show blank {} data. (using macOS 10.12.3 & Safari 10.0.3 (12602.4.8))

Can someone confirm that this issue is indeed present in Safari and that it is not just me experiencing this?

**Comment by thgreasi** [comment]
I think that IndexedDB on safari 10 should be resolved by #651. Can you test against the code in the current master (eg: by referencing the file in the dist folder).

**Comment by Narcwis** [comment]
I am currently using localforage 1.4.3 which seems to be same as the one in the current master.

**Comment by thgreasi** [comment]
OK , I will be waiting for your feedback.
Also, since you will be testing on safari, could you also provide some feedback on the jsfiddle example found in the last comment of #352.
PS: I hope I'm not asking too much from you.

**Comment by Narcwis** [comment]
I will be testing this tomorrow not only on Mac but also on iOS 9.3.3 and 10.0.2. Will give feedback as soon as I can. I've always benefited from open source projects so if I can do something to contribute, count me in.

**Comment by thgreasi** [comment]
@Narcwis any feedback on this?

**Comment by Narcwis** [comment]
@thgreasi The jsfiddle that you linked in the last comment of #352 works on all devices I have tested (iOS 9.3.3 and 10.0.2 and macOS Sierra with Safari 10.0.3). My problem as I said in my previous comment is that Safari does not seem to save blobs with localforage (I tried the files in the dist folder but to no success), I do not know if this is something related to the way Safari treats blobs or maybe something to do with localforage. I was hoping to narrow down the issue with your help. Do you have any suggestions?

**Comment by thgreasi** [comment]
@Narcwis thanks for the feedback on #352.
Can you try [this codepen](http://codepen.io/thgreasi/pen/vgVxzZ) that uses the current master, and report back the results and which storage method was used?

**Comment by Narcwis** [comment]
I am testing on macOS safari only for now when something works on that I will test on other devices as well. The codepen you linked works fine, saves an [object Blob] and reads an [object Blob] using localstorageWrapper. Saving a blob directly seems to be working but if you try to store the blob as an object's property that fails in Safari. I have taken the liberty of amending your codepen to reproduce the bug. Try [this codepen](http://codepen.io/narcwis/pen/GrYmRd) and let me know what you think.

**Comment by thgreasi** [comment]
>using localstorageWrapper

Oh, I forgot that 
Please use [this jsfiddle](https://jsfiddle.net/euhejqvh/) and right click in the result area and do a right-click => open in new tab.
If you can confirm that it works with indexed db, then we can close this issue.

Moreover, the nested blobs issue is already tracked in #380 and there is also PR #603 that resolves it, but it is under consideration because of the possible performance hit.

**Comment by thgreasi** [comment]
Please note that Safari <10.1 does not meet our requirements for IDB support (ref: #650, #651).
So for typical reasons I think that this one should be closed.
On the other hand I'm still interested for your findings, (in case that you are using 10.1),

**Comment by Narcwis** [comment]
The jsfiddle you linked returns, when opened in a new tab, using webSQLStorage not indexed db. (Safari 10.0.3)

About the nested blobs, the issue is not present in any other browser apart from Safari 10.0.3 (have yet to try 10.1). Regardless of the driver Safari won't save nested blobs, I will update to 10.1 and edit this comment with feedback. If 10.1 does not work will it be the case to open an issue for it? (I am assuming that #603 is already implemented in the latest version 1.4.3)

**Comment by thgreasi** [comment]
> I am assuming that #603 is already implemented in the latest version 1.4.3

No, this is still not merged, neither in v1.4.3 nor in master.

> If 10.1 does not work will it be the case to open an issue for it?

Regardless of the blob results, if 10.1 doesn't support IndexedDB on your tests, please report at #650.

**Comment by Narcwis** [comment]
Safari 10.2 uses asyncStorage when opened in new tab. 

The issue reproduced in my codepen with nested blobs also seems to be resolved but only when safari uses asyncstorage (opened in new tab), when using localstoragewrapper the issue persists. I am currently amending my project to see if I can get it working with asyncstorage.
